source/session.ts

```diff
@@ -669,28 +670,33 @@ export class Session {
    *   Return update or create promise.
    */
 
-  ensure<T extends Data = Data>(
+  ensure<
+    T extends
+      EntityTypeMap[keyof EntityTypeMap] = EntityTypeMap[keyof EntityTypeMap],
+  >(
     entityType: string,
     data: T,
     identifyingKeys: Array<keyof T> = [],
   ): Promise<T & Entity> {
     let keys = identifyingKeys as string[];
 
+    const anyData = data as any;
```

Comment:
I don't like this, but looking at it I can see that it gets a bit thorny to type the function better. We can add that as a later thing.

Reply:
Yeah that was exactly my thoughts too. It's definitely possible to strong-type it, but sometimes you have to go two steps forward and one step back I guess.